[DT][SVE] adjust tile sizes for mmt4d & disable transposition of narrow-N matmuls

egebeysel opened the pull request:
This PR builds upon #21304 and takes a step forward to adjust tile sizes in the existence of scalable inner tiles with `mmt4d`. This is step 2/N to enable data-tiling with SVE. There will be others to adjust tile sizes for packs/unpacks as well.

Also, since SVE requires the `N` dimension to be scalable - AFAIK because of vectorization unrolling on the inner `M` dimension and this currently not being supported for scalable `M` dims - disables transposition of narrow-N matmuls.

On another note, we currently walk the IR to retrieve the inner tile sizes of the `mmt4d`. This mechanism currently assumes that only `M` and `N` dimensions could be scalable, which holds for SME and SVE.
Although for future references, we could either find another mechanism to communicate the inner tile sizes chosen during materialization to the tile size selection OR extend this method to handle the inputs - and therefore infer the scalable K as well.

Note: materialization of SME is currently not supported.

egebeysel committed:
[DT][SVE] adjust tile sizes for mmt4d & disable transposition of
narrow-N types

Signed-off-by: Ege Beysel <beyselege@gmail.com>

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1848,8 +1850,104 @@ setRootConfig(mlir::FunctionOpInterface entryPointFn,
+  SmallVector<OpFoldResult> innerDims;
+  // Currently, the only init tensor of the mmt4d seems to be linalg.fill ops
+  // or the bindings. Otherwise, bail out.
+  if (auto fillOp = dyn_cast<linalg::FillOp>(destOp)) {
```
[nit] Document the code through early exits :)
```cpp
if (!isa<linalg::FillOp, tensor::EmptyOp>(destOp) return failure();
```

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1848,8 +1850,104 @@ setRootConfig(mlir::FunctionOpInterface entryPointFn,
 
+// Utility function to return the result inner tile sizes associated with the
+// mmt4d op. I.e. returns the M1 and N1 dimensions of the result shape
+// BxM0xN0xM1xN1. In case of scalable vectors, walks the IR and returns
```
Note, there are no vectors at this level. Also, could you document what happens for static sizes?
```suggestion
// BxM0xN0xM1xN1. In case of scalable dimensions, walks the IR and returns
```

banach-space commented on compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:
```diff
@@ -1592,6 +1592,52 @@ getDefaultVscaleRange(IREE::HAL::ExecutableTargetAttr targetAttr) {
+  SmallVector<int64_t> tileSizes(mixedInnerTiles.size(), 0);
+  IREE::Codegen::ScalableTileFlags scalableFlags(mixedInnerTiles.size(), false);
+  for (unsigned pos = 0; pos < mixedInnerTiles.size(); ++pos) {
+    if (auto innerTileVal = dyn_cast<Value>(mixedInnerTiles[pos])) {
+      FailureOr<int64_t> innerTile =
+          getStaticPartOfScalableTileSize(innerTileVal.getDefiningOp());
+      if (failed(innerTile)) {
+        LDBG() << "Found non-scalable dynamic inner tile!";
+        return failure();
+      }
+      tileSizes[pos] = innerTile.value();
+      scalableFlags[pos] = true;
+      continue;
+    }
+    std::optional<int64_t> innerTile =
+        getConstantIntValue(mixedInnerTiles[pos]);
+    if (!innerTile.has_value()) {
+      LDBG() << "Error while inferring static inner tile size!";
+      return failure();
+    }
```
[nit] The two cases were not immediately obvious to me. Perhaps add comments?

```suggestion
    // Scalable tile?
    if (auto innerTileVal = dyn_cast<Value>(mixedInnerTiles[pos])) {
      FailureOr<int64_t> innerTile =
          getStaticPartOfScalableTileSize(innerTileVal.getDefiningOp());
      if (failed(innerTile)) {
        LDBG() << "Found non-scalable dynamic inner tile!";
        return failure();
      }
      tileSizes[pos] = innerTile.value();
      scalableFlags[pos] = true;
      continue;
    }
    // Constant tile?
    std::optional<int64_t> innerTile =
        getConstantIntValue(mixedInnerTiles[pos]);
    if (!innerTile.has_value()) {
      LDBG() << "Error while inferring static inner tile size!";
      return failure();
    }
```

banach-space commented:
Thanks!

> Note: materialization of SME is currently not supported - though the logic to adjust the tile sizes is exactly the same as SVE - so this is also included. To enable SME materialization, we should experiment with inner tile sizes and add these accordingly.

Could you include a note near the implementation? (e.g. "TODO: Enable materialization and test").

> Also, since SVE requires the N dimension to be scalable - AFAIK because of vectorization unrolling on the inner M dimension and this currently not being supported for scalable M dims - disables transposition of narrow-N matmuls.

Not sure about transposes (will ask offline), but with regard to M vs N dim, we have basically focused on making the `N` dim scalable as the canonical form of "scalable vectorisation" and one that follows:
* https://developer.arm.com/-/media/developer/products/software-tools/hpc/White%20papers/arm-scalable-vector-extensions-and-application-to-machine-learning.pdf?revision=510ee340-fce1-4fd8-bad6-bade674620a5

Making the `M` dim scalable should be possible, but we haven't really tried yet :) (there have been other priorities) Hope this clarifies a bit.

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1848,8 +1850,104 @@ setRootConfig(mlir::FunctionOpInterface entryPointFn,
+      ShapedType::isDynamic(N0)) {
+    int64_t innerTileSize;
+    bool scalableFlag;
+    if (failed(scalableInnerTilesAndFlags)) {
+      // For SVE, the N dimension is currently chosen as 8 - unless we have RHS
+      // with i4 type and target does not have "+dotprod" or "+i8mm" features.
+      // Currently, we hard-code that over here as a fallback option.
+      LDBG() << "Using default vector parallel tile sizes for the inner M dim!";
+      auto rhsElemType = rhsType.getElementType();
+      innerTileSize = rhsElemType.isSignlessInteger(4) &&
+                              !(hasFeature(targetConfig, "+i8mm") ||
+                                hasFeature(targetConfig, "+dotprod"))
+                          ? 16
+                          : 8;
+      scalableFlag = true;
```
Are we testing this block?

egebeysel replied:
I'll add a test for this! Though this would be the case if we have an inner tile size that doesn't look like `vscale * staticSize`, which should not be the case IMO. I'll add one in which the inner tile size is passed as a function argument.

egebeysel replied:
I deleted this block and also disabled SME for now, TODO added for later.

banach-space reviewed (commented):
Great!

I've quickly skimmed through before being (mostly) offline for a week.

hanhanW commented:
cc @bjacob who may be interested in this.

egebeysel committed:
address reviews & add comments

Signed-off-by: Ege Beysel <beyselege@gmail.com>

egebeysel committed:
fixup test - remove module

Signed-off-by: Ege Beysel <beyselege@gmail.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/ExternalInterfaces/CPUEncodingExternalModels.cpp:
```diff
@@ -326,6 +330,10 @@ FailureOr<Operation *> lowerContractionOpWithEncoding(
   }
 
   bool transpose = isNarrowNResult(resultEncoding);
+  // Do not transpose in case we have scalable tiles.
+  transpose &= llvm::none_of(
+      encodingInfo.scalableTiles.value_or(IREE::Codegen::ScalableTileFlags{}),
+      [](bool flag) { return flag; });
```
This is exposing the logic about arch to local function. The better fix to me is querying such information from `CPUEncodingResolverAttr` and pass it to local functions.

I think it is better to set the field in the configuration.

https://github.com/iree-org/iree/blob/337c8aaf544053ac512bb7c023c572f8e2d51e7e/compiler/src/iree/compiler/Codegen/ExternalInterfaces/CPUEncodingExternalModels.cpp#L739-L752

hanhanW replied:
Sorry for the bad suggestion. I talked to Benoit and I understood the idea of transposition better now. You can ignore this comment and leave as what it is; I'll just do an easy and better fix later on.

egebeysel replied:
No problem, I haven't had time to address the comments yet :)

Could you also enlighten me about them if you don't mind, thanks!

hanhanW commented on compiler/src/iree/compiler/Codegen/ExternalInterfaces/CPUEncodingExternalModels.cpp:
```diff
@@ -701,14 +710,16 @@ struct CPUEncodingPackedLayoutMaterializerAttr
     if (succeeded(scalableFlags)) {
       info.scalableTiles = std::move(scalableFlags);
     }
+    if (IREE::Encoding::isNarrowNResult(encoding) &&
+        llvm::none_of(info.scalableTiles.value_or(Codegen::ScalableTileFlags{}),
+                      [](bool flag) { return flag; })) {
+      transposeInPlace(info);
+    }
```
Then we can query the field in this function as well.

hanhanW reviewed (commented):
I haven't reviewed changes in `KernelDispatch.cpp` yet, and I'm still working on it. I'll try to do it later today or tomorrow. Feel free to wait for the next round of review comments before you make changes.

hanhanW commented on compiler/src/iree/compiler/Codegen/Utils/Utils.h:
```diff
@@ -275,9 +275,35 @@ void sinkOpsInCFG(const SmallVector<Operation *> &allocs,
+using SizesAndScalableFlags =
+    std::pair<SmallVector<int64_t>, SmallVector<bool>>;
+
+// This utility function returns the static part of scalable inner tile sizes.
+// These are - as of now always and probably should always be - arith.muli ops
+// with `vector.vscale` on one side and an `arith.constant` on the other. It
+// returns the constant if found.
+FailureOr<int64_t> getStaticPartOfScalableTileSize(Operation *op);
```
Should this be a local function given the only user is `getScalableTileSizesAndFlags`?

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1852,8 +1854,107 @@ setRootConfig(mlir::FunctionOpInterface entryPointFn,
+// Adjusts the tile sizes and scalable flags for SVE and SME in-place.
+// Returns failure if it cannot adjust tile sizes.
+static LogicalResult adjustVectorSizesForScalableVectorization(
+    linalg::LinalgOp op, DictionaryAttr targetConfig, int64_t M0, int64_t N0,
```
style nit: use `m0` and `n0`. We follow camel naming convention.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1852,8 +1854,107 @@ setRootConfig(mlir::FunctionOpInterface entryPointFn,
+    SmallVector<int64_t> &vecTileSizes,
+    IREE::Codegen::ScalableTileFlags &vecScalableTileFlags) {
+  int64_t mmt4dDimBase = isa<linalg::BatchMmt4DOp>(op) ? 1 : 0;
+  ShapedType rhsType = cast<ShapedType>(op.getDpsInputs()[1].getType());
```
auto

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1913,11 +2014,17 @@ getMmt4dLoweringConfig(linalg::LinalgOp op) {
-
+  IREE::Codegen::ScalableTileFlags vecScalableTileFlags(mmt4dDimBase + 6,
+                                                        false);
+  if (!(targetConfig && isAArch64(targetConfig) &&
+        isScalableVectorizationEnabled() &&
+        succeeded(adjustVectorSizesForScalableVectorization(
+            op, targetConfig, M0, N0, vecTileSizes, vecScalableTileFlags)))) {
+    limitVectorTileSizes(op, vecTileSizes);
+  }
```
This is too hard to follow.. Can we just create different functions for scalable and non-scalable cases for getting the vector sizes?

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1852,8 +1854,107 @@ setRootConfig(mlir::FunctionOpInterface entryPointFn,
+  SmallVector<OpFoldResult> innerDims;
+  // Currently, the only init tensor of the mmt4d seems to be linalg.fill ops
+  // or the bindings. Otherwise, bail out.
+  if (!isa<linalg::FillOp, IREE::TensorExt::DispatchTensorLoadOp>(destOp)) {
```
The better solution to me is implementing `ShapeAwareOpInterface` for `tensor.empty` op and track the list of `linalg.fill` until it reaches the interface op, but I'm unsure if it opens a can of worms or not. I'll try to implement it and see if it works.

banach-space reviewed (commented):
Thanks for the updates!

egebeysel commented:
I will address all of the reviews soon - I've just been caught up in something else and didn't have time, sorry!

egebeysel committed:
adress comments

Signed-off-by: Ege Beysel <beyselege@gmail.com>

egebeysel committed:
rebase and reformat tests

Signed-off-by: Ege Beysel <beyselege@gmail.com>

egebeysel commented:
I'm sorry I couldn't do anything for a long while on this PR, had to work on a WIP branch for a while :) I addressed your comments, reformatted the tests a bit and rebased the PR. Please take another look! @banach-space @hanhanW

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1873,8 +1875,84 @@ setContractionRootConfig(mlir::FunctionOpInterface entryPointFn,
+// Returns failure if it cannot adjust tile sizes.
+static bool adjustVectorSizesForScalableVectorization(
+    linalg::LinalgOp op, DictionaryAttr targetConfig, int64_t m0, int64_t n0,
+    SmallVector<int64_t> &vecTileSizes,
```
SmallVectorImpl

hanhanW replied:
Sorry for not being clear in the first place. I meant to use `SmallVectorImpl` instead of `SmallVector`. `SmallVector<T>` infers the `N` for `SmallVector<T, N>` and they all inherits from `SmallVectorImpl<T>`. Using `SmallVectorImpl<T>&` allows your function to accept any `SmallVector` instance, regardless of its specific static inline capacity N. This makes your function more general and reusable.

egebeysel committed:
address comments

Signed-off-by: Ege Beysel <beyselege@gmail.com>
